xdc-plugin/xdc.cc

```diff
@@ -0,0 +1,172 @@
+		log("\n");
+		log("   get_ports \n");
+		log("\n");
+		log("Get matching ports\n");
+		log("\n");
+		log("Print the output to stdout too. This is useful when all Yosys is executed\n");
+		log("\n");
+	}
+	void execute(std::vector<std::string> args, RTLIL::Design*) YS_OVERRIDE
+	{
+		std::string text;
+		for (auto& arg : args) {
+			text += arg + ' ';
+		}
+		if (!text.empty()) text.resize(text.size()-1);
+		log("%s\n", text.c_str());
+	}
+} GetPorts;
+
+struct GetIOBanks : public Pass {
+	GetIOBanks() : Pass("get_iobanks", "Set IO Bank number") {}
+	void help() YS_OVERRIDE
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("   get_iobanks \n");
+		log("\n");
+		log("Get IO Bank number\n");
+		log("\n");
+	}
+	void execute(std::vector<std::string> args, RTLIL::Design* ) YS_OVERRIDE
+	{
+		if (args.size() != 2) {
+			log("Incorrect number of arguments. %zu instead of 1", args.size());
+			return;
+		}
+		current_iobank = std::atoi(args[1].c_str());
+		// Check if the part has the specified bank
+		BankTilesMap bank_tiles(get_bank_tiles());
+		if (bank_tiles.count(current_iobank) == 0) {
+			log("get_iobanks: Incorrect bank number: %d\n", current_iobank);
+			current_iobank = 0;
+		}
+	}
+} GetIOBanks;
+
+struct SetProperty : public Pass {
+	SetProperty() : Pass("set_property", "Set a given property") {}
+	void help() YS_OVERRIDE
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("    set_property PROPERTY VALUE OBJECT\n");
+		log("\n");
+		log("Set the given property to the specified value on an object\n");
+		log("\n");
+	}
+	void execute(std::vector<std::string> args, RTLIL::Design* design) YS_OVERRIDE
+	{
+		if (design->top_module() == nullptr) {
+			log("No top module detected\n");
+			return;
+		}
+
+		std::string option(args[1]);
+		if (set_property_options_map.count(option) == 0) {
+			log("set_property: %s option is currently not supported\n", option.c_str());
+			return;
+		}
+
+		switch (set_property_options_map[option]) {
+			case INTERNAL_VREF:
+				process_vref(std::vector<std::string>(args.begin() + 2, args.end()), design);
+				break;
+			default:
+				assert(false);
+		}
+	}
+	void process_vref(std::vector<std::string> args, RTLIL::Design* design)
+       	{
+		if (args.size() != 2) {
+			log("set_property INTERNAL_VREF: Incorrect number of arguments: %zu\n", args.size());
+			return;
+		}
+
+		if (current_iobank == 0) {
+			log("set_property INTERNAL_VREF: No valid bank set. Use get_iobanks.\n");
+			return;
+		}
+
+		int internal_vref = 1000 * std::atof(args[0].c_str());
+		if (internal_vref != 600 &&
+				internal_vref != 675 &&
+				internal_vref != 750 &&
+				internal_vref != 900) {
+			log("set_property INTERNAL_VREF: Incorrect INTERNAL_VREF value\n");
+			return;
+		}
+
+		// Create a new BANK module if it hasn't been created so far
```

Comment:
Can these be in Verilog file rather than hard coded into this C++ file?

Reply:
Done